src/systems/environmental_sensor_system/EnvironmentalSensorSystem.cc

```diff
@@ -0,0 +1,377 @@
+        this->session.value(), std::chrono::duration<double>(_now).count());
+
+    if (!this->session.has_value()) return false;
+
+    gz::msgs::Double msg;
+    *msg.mutable_header()->mutable_stamp() = gz::msgs::Convert(_now);
+    auto frame = msg.mutable_header()->add_data();
+    frame->set_key("frame_id");
+    frame->add_value((this->frame_id == "") ? this->Name() : this->frame_id);
+    auto data = this->gridField->frame[this->field].LookUp(
+      this->session.value(), this->position);
+    if (!data.has_value())
+    {
+      gzwarn << "Failed to acquire value perhaps out of field?\n";
+      return false;
+    }
+    msg.set_data(data.value());
+    // TODO(anyone) Add sensor noise.
+    this->pub.Publish(msg);
+    return true;
+  }
+
+  ////////////////////////////////////////////////////////////////
+  /// \brief Attempts to set a data table.
+  /// \param[in] _data - The data table
+  /// \param[in] _curr_time - The current time.
+  public: void SetDataTable(
+    const components::Environment* _data,
+    const std::chrono::steady_clock::duration &_curr_time)
+  {
+    gzdbg << "Setting new data table\n";
+    auto data = _data->Data();
+    if(!data->frame.Has(this->field))
+    {
+      gzwarn << "Environmental sensor could not find field "
+        << this->field << "\n";
+      this->ready = false;
+      return;
+    }
+
+    this->gridField = data;
+    this->session = this->gridField->frame[this->field].CreateSession();
+    if (!this->gridField->staticTime)
+    {
+      this->session = this->gridField->frame[this->field].StepTo(
+        *this->session,
+        std::chrono::duration<double>(_curr_time).count());
+    }
+    this->ready = true;
+
+    if(!this->session.has_value())
+    {
+      gzerr << "Exceeded time stamp." << std::endl;
+    }
+  }
+
+  ////////////////////////////////////////////////////////////////
+  /// \brief Updates the position of the sensor at each time step
+  /// to track a given entity,
+  /// \param[in] - The sensor entity.
+  /// \param[in] - The ECM to use for tracking.
+  /// \returns True if successful, false if not.
+  public: bool UpdatePosition(
+    const Entity _entity,
+    const EntityComponentManager& _ecm)
+  {
+    if (!this->ready) return false;
+
+    const auto position = worldPose(_entity, _ecm).Pos();
+    if (this->gridField->reference != math::SphericalCoordinates::LOCAL2)
+    {
+        auto origin =
+          _ecm.Component<components::SphericalCoordinates>(worldEntity(_ecm));
+        if (!origin)
+        {
+          gzerr << "World has no spherical coordinates,"
+              <<" but data was loaded with spherical reference plane" 
+              << std::endl;
+          return false;
+        }
+        this->position = origin->Data().PositionTransform(
+            position, math::SphericalCoordinates::LOCAL2,
+            this->gridField->reference);
+        if (this->gridField->units ==
+          components::EnvironmentalData::ReferenceUnits::DEGREES)
```

Comment:
@arjo129 hmm, consider checking for the reference type before assuming this conversion applies.